Kernel/Bus/USB/xHCI/DeviceTreexHCIController.cpp

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2025, Sönke Holz <sholz8530@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/Bus/USB/USBManagement.h>
+#include <Kernel/Bus/USB/xHCI/DeviceTreexHCIController.h>
+#include <Kernel/Bus/USB/xHCI/xHCIInterrupter.h>
+#include <Kernel/Firmware/DeviceTree/DeviceTree.h>
+#include <Kernel/Firmware/DeviceTree/Driver.h>
+#include <Kernel/Firmware/DeviceTree/Management.h>
+
+namespace Kernel::USB {
+
+ErrorOr<NonnullLockRefPtr<DeviceTreexHCIController>> DeviceTreexHCIController::try_to_initialize(DeviceTree::Device::Resource registers_resource, StringView node_name, size_t interrupt_number)
+{
+    auto registers_mapping = TRY(Memory::map_typed_writable<u8>(registers_resource.paddr));
+
+    auto controller = TRY(adopt_nonnull_lock_ref_or_enomem(new (nothrow) DeviceTreexHCIController(move(registers_mapping), node_name, interrupt_number)));
+    TRY(controller->initialize());
+    return controller;
+}
+
+UNMAP_AFTER_INIT DeviceTreexHCIController::DeviceTreexHCIController(Memory::TypedMapping<u8> registers_mapping, StringView node_name, size_t interrupt_number)
+    : xHCIController(move(registers_mapping))
+    , m_node_name(node_name)
+    , m_interrupt_number(interrupt_number)
+{
+}
+
+ErrorOr<NonnullOwnPtr<GenericInterruptHandler>> DeviceTreexHCIController::create_interrupter(u16 interrupter_id)
+{
+    return TRY(xHCIDeviceTreeInterrupter::create(*this, m_interrupt_number, interrupter_id));
+}
+
+static constinit Array const compatibles_array = {
+    "generic-xhci"sv,
+};
+
+DEVICETREE_DRIVER(DeviceTreexHCIControllerDriver, compatibles_array);
+
+// https://www.kernel.org/doc/Documentation/devicetree/bindings/usb/generic-xhci.yaml
+ErrorOr<void> DeviceTreexHCIControllerDriver::probe(DeviceTree::Device const& device, StringView) const
+{
+    auto registers_resource = TRY(device.get_resource(0));
+
+    auto const interrupt = TRY(device.node().interrupts(DeviceTree::get()))[0];
+
+    // FIXME: Don't depend on a specific interrupt descriptor format and implement proper devicetree interrupt mapping/translation.
+    if (!interrupt.domain_root->is_compatible_with("arm,gic-400"sv) && !interrupt.domain_root->is_compatible_with("arm,cortex-a15-gic"sv))
```

Comment:
Later on we should also check if the device is not discretely a PCI device, IIRC some devices might be connected over PCI, but are listed in a simple-bus

Reply:
Yes, that is actually the case on the Pi 5, where the Synopsys xHCs are in the RP1 node.

This code should already throw an error if the parent node is a PCI controller, because the address translation code currently returns ENOTSUP if it encounters a PCI bus.
The interrupt getter would also fail, because PCI busses use interrupt nexuses (which we don't support yet) to describe the PCI interrupt mapping.